lib/THC/THCThreadLocal.c

```diff
@@ -0,0 +1,43 @@
+#include "THCThreadLocal.h"
+#include "THCGeneral.h"
+
+
+THCThreadLocal THCThreadLocal_alloc()
+{
+#ifndef _WIN32
+  pthread_key_t key;
+  THAssert(pthread_key_create(&key, NULL) == 0);
+  return key;
+#else
+  DWORD key = TlsAlloc();
+  THAssert(key != TLS_OUT_OF_INDEXES);
+  return key;
+#endif
+}
+
+void THCThreadLocal_free(THCThreadLocal local)
+{
+#ifndef _WIN32
+  THAssert(pthread_key_delete(local) == 0);
+#else
+  THAssert(TlsFree(local));
+#endif
+}
+
+void* THCThreadLocal_get(THCThreadLocal local)
```

Comment:
We're not using C++ here or in THCGeneral.c. Adding templates doesn't seem like it'll **reduce** code complication, especially since you can't really specialize the POSIX or Windows thread-local API. They both take `void*`, not arbitrary types.